pkg/services/pluginsintegration/clientmiddleware/caching_middleware.go

```diff
@@ -0,0 +1,142 @@
+			caching: cachingService,
+			log:     log,
+		}
+	})
+}
+
+type CachingMiddleware struct {
+	next    plugins.Client
+	caching caching.CachingService
+	log     log.Logger
+}
+
+// QueryData receives a data request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the queries as usual, then write the response to the cache.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	if req == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.QueryData(ctx, req)
+	}
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the query cache if enabled
+	hit, cr := m.caching.HandleQueryRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
+			QueryRequestHistogram.With(prometheus.Labels{
+				"datasource_type": req.PluginContext.DataSourceInstanceSettings.Type,
+				"cache":           ch,
+				"query_type":      getQueryType(reqCtx),
+			}).Observe(time.Since(start).Seconds())
+		}
+	}()
+
+	// Cache hit; return the response
+	if hit {
+		return cr.Response, nil
+	}
+
+	// Cache miss; do the actual queries
+	resp, err := m.next.QueryData(ctx, req)
+
+	// Update the query cache with the result for this metrics request
+	if err == nil && cr.UpdateCacheFn != nil {
+		cr.UpdateCacheFn(ctx, resp)
+	}
+
+	return resp, err
+}
+
+// CallResource receives a resource request and attempts to access results already stored in the cache for that request.
+// If data is found, it will return it immediately. Otherwise, it will perform the request as usual. The caller of CallResource is expected to explicitly update the cache with any responses.
+// If the cache service is implemented, we capture the request duration as a metric. The service is expected to write any response headers.
+func (m *CachingMiddleware) CallResource(ctx context.Context, req *backend.CallResourceRequest, sender backend.CallResourceResponseSender) error {
+	if req == nil {
+		return m.next.CallResource(ctx, req, sender)
+	}
+
+	reqCtx := contexthandler.FromContext(ctx)
+	if reqCtx == nil {
+		return m.next.CallResource(ctx, req, sender)
+	}
+
+	// time how long this request takes
+	start := time.Now()
+
+	// First look in the resource cache if enabled
+	hit, resp := m.caching.HandleResourceRequest(ctx, req)
+
+	defer func() {
+		// record request duration if caching was used
+		if ch := reqCtx.Resp.Header().Get(caching.XCacheHeader); ch != "" {
+			ResourceRequestHistogram.With(prometheus.Labels{
+				"datasource_type": req.PluginContext.DataSourceInstanceSettings.Type,
```

Comment:
`req.PluginContext.DataSourceInstanceSettings` would be nil if request is made for an app plugin.

```suggestion
				"plugin_id": req.PluginContext.PluginID,
```